crates/router/src/connector/nmi.rs

```diff
@@ -782,24 +792,115 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
+
+        let regex_pattern = r"t=(.*),s=(.*)";
+
+        if let Some(captures) = Regex::new(regex_pattern)
+            .into_report()
+            .change_context(errors::ConnectorError::WebhookSignatureNotFound)?
+            .captures(sig_header)
+        {
+            let nonce = captures
+                .get(0)
+                .ok_or(errors::ConnectorError::WebhookSignatureNotFound)
+                .into_report()?
+                .as_str();
+
+            let message = format!("{}.{}", nonce, String::from_utf8_lossy(request.body));
+
+            return Ok(message.into_bytes());
+        }
+        Err(errors::ConnectorError::WebhookSignatureNotFound).into_report()
+    }
+
+    fn get_webhook_object_reference_id(
+        &self,
+        request: &api::IncomingWebhookRequestDetails<'_>,
     ) -> CustomResult<api_models::webhooks::ObjectReferenceId, errors::ConnectorError> {
-        Err(errors::ConnectorError::WebhooksNotImplemented).into_report()
+        let reference_body: nmi::NmiWebhookObjectReference = request
+            .body
+            .parse_struct("nmi NmiWebhookObjectReference")
+            .change_context(errors::ConnectorError::WebhookResourceObjectNotFound)?;
+
+        let object_reference_id = match reference_body.event_body.action.action_type {
+            nmi::NmiActionType::Sale => api_models::webhooks::ObjectReferenceId::PaymentId(
+                api_models::payments::PaymentIdType::PaymentAttemptId(
+                    reference_body.event_body.order_id,
+                ),
+            ),
+            nmi::NmiActionType::Refund => api_models::webhooks::ObjectReferenceId::RefundId(
+                api_models::webhooks::RefundIdType::RefundId(reference_body.event_body.order_id),
+            ),
+            _ => Err(errors::ConnectorError::WebhooksNotImplemented).into_report()?,
+        };
+
+        Ok(object_reference_id)
     }
 
     fn get_webhook_event_type(
         &self,
-        _request: &api::IncomingWebhookRequestDetails<'_>,
+        request: &api::IncomingWebhookRequestDetails<'_>,
     ) -> CustomResult<api::IncomingWebhookEvent, errors::ConnectorError> {
-        Ok(api::IncomingWebhookEvent::EventNotSupported)
+        let event_type_body: nmi::NmiWebhookEventBody = request
+            .body
+            .parse_struct("nmi NmiWebhookEventType")
+            .change_context(errors::ConnectorError::WebhookResourceObjectNotFound)?;
+
+        Ok(api::IncomingWebhookEvent::foreign_from(
+            event_type_body.event_type,
+        ))
     }
 
     fn get_webhook_resource_object(
         &self,
-        _request: &api::IncomingWebhookRequestDetails<'_>,
+        request: &api::IncomingWebhookRequestDetails<'_>,
     ) -> CustomResult<Box<dyn masking::ErasedMaskSerialize>, errors::ConnectorError> {
-        Err(errors::ConnectorError::WebhooksNotImplemented).into_report()
+        let webhook_body: nmi::NmiWebhookBody = request
```

Comment:
Since, this object will be consumed by PSync handle_response, let's have try_from webhook response object to  PSync response object for payments and in case of refund return the object as it is. This will ensure that any changes in PSync will not break the incoming webhook flow